### PR TITLE
Restore accessibility to all profiles when default profile has APIML token authentication

### PR DESCRIPTION
**Pull request by likhithanimma1**
## Proposed changes

   Solves #2111 
   Fixes the issue of accessibility of all the profiles when the default profile is using the APIML token authentication

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 1]
```diff
@@ -539,11 +539,10 @@ export class ProfilesCache {
...
 
     private shouldRemoveTokenFromProfile(profile: zowe.imperative.IProfileLoaded, baseProfile: zowe.imperative.IProfileLoaded): boolean {
-        return (baseProfile?.profile?.host &&
-            baseProfile?.profile?.port &&
+        return ((baseProfile?.profile?.host || baseProfile?.profile?.port) &&
             profile?.profile?.host &&
             profile?.profile?.port &&
-            (baseProfile?.profile.host !== profile?.profile.host || baseProfile?.profile.port !== profile?.profile.port) &&
+            (baseProfile?.profile.host !== profile?.profile.host || (profile?.profile.user && profile?.profile.password)) &&
             profile?.profile.tokenType?.startsWith(zowe.imperative.SessConstants.TOKEN_TYPE_APIML)) as boolean;
```
Comment (traeok) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:546]: There are a lot of factors in this boolean logic here 😅 

Would it make sense to move this line to the top so we return false before doing the other checks? If the profile doesn't have a token type, we can immediately return false.

**Comment by JillieBeanSim** [comment]
should this be included in 2.15.0?

**Review by JillieBeanSim — APPROVED** [review]

**Comment by zFernand0** [comment]
> should this be included in 2.15.0?

we could include it in 2.15.0

---

Also, I haven't approved the PR because I can still see the issue (when tested in the vNext version of this PR).
I believe there was a bit of info missing in the reproduction steps.
Where are the credentials coming from? (base or service)

- If the credentials were initially in the base profile, and you did a `zowe auth login apiml`, 
  - You will get a new base profile created (with the word `_apiml` as a suffix)
  - This new `base_apiml` profile gets assigned as the default
  - This will cause other service profiles to break (because there are no credentials in the default base profile anymore)
  - I was able to reproduce that error in this PR

If you want to avoid this from happening:
 - You must add the credentials on the service profile (or somewhere in the nested structure).
 - This works fine in the version out in the marketplace

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 2]
```diff
@@ -539,11 +539,10 @@ export class ProfilesCache {
     }
 
     private shouldRemoveTokenFromProfile(profile: zowe.imperative.IProfileLoaded, baseProfile: zowe.imperative.IProfileLoaded): boolean {
-        return (baseProfile?.profile?.host &&
-            baseProfile?.profile?.port &&
+        return ((baseProfile?.profile?.host || baseProfile?.profile?.port) &&
             profile?.profile?.host &&
             profile?.profile?.port &&
-            (baseProfile?.profile.host !== profile?.profile.host || baseProfile?.profile.port !== profile?.profile.port) &&
```
Comment (traeok) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:546]: Can we add this check back for comparing the base profile's port with the port of the profile that was passed in? I like the added check for user and password on the profile, but I think keeping the port check will help narrow the opportunity for error until something more comprehensive is incorporated into Zowe CLI/Imperative.

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 3]
```diff
@@ -539,11 +539,10 @@ export class ProfilesCache {
...
+            (baseProfile?.profile.host !== profile?.profile.host || (profile?.profile.user && profile?.profile.password)) &&
```
Comment (zFernand0) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:545]: Also, it might be good to keep the `port` check as well 😋

**Review by traeok — CHANGES_REQUESTED** [review]

**Comment by traeok** [comment]
> If you want to avoid this from happening:
> 
> * You must add the credentials on the service profile (or somewhere in the nested structure).
> * This works fine in the version out in the marketplace

I had tested this PR under the assumption that credentials were stored on the service profile. However, the details for #2111 do not mention whether credentials were stored on the service or base profile.

Considering that its unclear of what authentication method to use for a profile in that scenario, I think it would be best to accept this fix until a more comprehensive solution is introduced on the Zowe CLI/Imperative side (such as profile linking). However, I think we should also keep the port check until this solution is introduced.

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts (hunk 3 above)
Comment (t1m0thyj) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:545]: Agreed, it might be good to keep the `port` check here as well as on line 542

**Review by traeok — APPROVED** [review]
LGTM, thanks @likhithanimma1 for the bugfix! 🎉

**Review by JillieBeanSim — APPROVED** [review]

**Pull request merged**